# CSC 593

## Week 7

### Merge Errors

#### Not resolved automatically
`git checkout --ours PATH/FILE`

#### Resolved automatically (vim)
`:wq`

### Pandas 2

In [18]:
import numpy as np
import pandas as pd

The BRFSS (Behavioral Risk Factor Surveillance System) data is too big to put into Github. This cell downloads it from the CDC's website and unzips it into your `data` folder.

In [34]:
#Setup for examples.
from urllib.request import urlretrieve
import zipfile
from pathlib import Path

zf = '../data/brfss/LLCP2018ASC.zip'
if not Path(zf).exists():
    Path('../data/brfss').mkdir(exist_ok=True)

    urlretrieve('https://www.cdc.gov/brfss/annual_data/2018/pdf/overview-2018-508.pdf', '../data/brfss/overview-2018-508.pdf')
    urlretrieve('https://www.cdc.gov/brfss/annual_data/2018/pdf/codebook18_llcp-v2-508.pdf', '../data/brfss/codebook18_llcp-v2-508.pdf')
    
    urlretrieve('https://www.cdc.gov/brfss/annual_data/2018/files/LLCP2018ASC.zip', zf)

fwff = '../data/brfss/LLCP2018.ASC '
if not Path(fwff).exists():
    with zipfile.ZipFile(zf) as z:
        z.extractall('../data/brfss')

Load the BRFSS data and set a couple of data types explicitly. (More supported data types are listed at https://docs.scipy.org/doc/numpy/user/basics.types.html)

In [35]:
names= ['state', 'imonth', 'iday', 
        'iyear', 'dispcode','genhlth', 
        'physhlth',
        'menthlth', 'poorhlth', 'hlthpln1',
        'persdoc2', 'medcost', 'checkup1',
        'WEIGHT2', 'HEIGHT3']
cols = [
    (1, 3),
    (18, 20),
    (20, 22),
    (22, 27),
    (31, 35),
    (89, 90),
    (90, 92),
    (92, 94),
    (94, 96),
    (96, 97),
    (97, 98),
    (98, 99),
    (99, 100),
    (176, 180),
    (180, 184)
]
types= {A
    'WEIGHT2': str, 
    'HEIGHT3': str,
}
brfss = pd.read_fwf(fwff, 
                    names=names,
                    colspecs=cols,
                    dtype=types)

#### Searching

In [67]:
#Get an individual column.
brfss['WEIGHT2']

0         0130
1         0200
2         0142
3         0190
5         0172
6         0150
7         0130
8         0205
9         0151
10        0199
11        0170
12        0160
13        0200
14        0285
15        0217
16        0230
17        0250
18        0155
19        0138
20        0240
21        0240
22        0185
23        0165
24        0131
25        0119
26        0180
27        0169
28        0128
29        0218
30        0148
          ... 
437404    0108
437405    0177
437407    0320
437408    0125
437409    0158
437410    0155
437411    0150
437412    0138
437413    0135
437414    0178
437415    0126
437416    0115
437417    0183
437418    0153
437419    0138
437420    0260
437421    0160
437422    0200
437423    0126
437424    0340
437425    0298
437426    0160
437427    0160
437428    0163
437429    0172
437430    0170
437431    0098
437433    0156
437434    0137
437435    0192
Name: WEIGHT2, Length: 404523, dtype: object

In [68]:
#For multiple columns, use a list as a subscript.
brfss[['WEIGHT2', 'HEIGHT3']]

WEIGHT2 HEIGHT3
0         0130    0504
1         0200    0505
2         0142    0410
3         0190    0510
5         0172    0502
6         0150    0505
7         0130    0505
8         0205    0601
9         0151    0503
10        0199    0600
11        0170    0503
12        0160    0503
13        0200    0504
14        0285    0511
15        0217    0506
16        0230    0509
17        0250    0601
18        0155    0504
19        0138    0504
20        0240    0507
21        0240    0504
22        0185    0505
23        0165    0509
24        0131    0504
25        0119    0502
26        0180    0505
27        0169    0506
28        0128    0506
29        0218    0603
30        0148    0502
...        ...     ...
437404    0108    0503
437405    0177    0503
437407    0320    0603
437408    0125    0501
437409    0158    0505
437410    0155    0502
437411    0150    0506
437412    0138    0502
437413    0135    0506
437414    0178    0506
437415    0126    0506
437416    0115    0411
437417    0183    0503
437418    0153    0504
437419    0138    0500
437420    0260    0503
437421    0160    0507
437422    0200    0601
437423    0126    0500
437424    0340    0507
437425    0298    0600
437426    0160    0508
437427    0160    0506
437428    0163    0502
437429    0172    0508
437430    0170    0503
437431    0098    0502
437433    0156    0500
437434    0137    0501
437435    0192    0510

[404523 rows x 2 columns]

The `loc()` and `iloc()` methods (see the table on p. 144-5 of *Python for Data Analysis*)

In [69]:
#Get the first row.
brfss.loc[0]

state          1
imonth         1
iday           5
iyear       2018
dispcode    1100
genhlth        2
physhlth      30
menthlth      88
poorhlth      30
hlthpln1       1
persdoc2       1
medcost        2
checkup1       1
WEIGHT2     0130
HEIGHT3     0504
wtunit         0
wt           130
wtlbs        130
htunit         0
htinches     504
Name: 0, dtype: object

In [70]:
#Get WEIGHT2 from the third row.
brfss.loc[2, 'WEIGHT2']

'0142'

In [71]:
#Same thing, but using the integer index instead of the column name.
brfss.iloc[2,-2]

0

In [72]:
brfss.at[2, 'WEIGHT2']

'0142'

In [73]:
brfss.iat[2, -2]

0

In [74]:
brfss.loc[:100, ['HEIGHT3', 'WEIGHT2']]

HEIGHT3 WEIGHT2
0      0504    0130
1      0505    0200
2      0410    0142
3      0510    0190
5      0502    0172
6      0505    0150
7      0505    0130
8      0601    0205
9      0503    0151
10     0600    0199
11     0503    0170
12     0503    0160
13     0504    0200
14     0511    0285
15     0506    0217
16     0509    0230
17     0601    0250
18     0504    0155
19     0504    0138
20     0507    0240
21     0504    0240
22     0505    0185
23     0509    0165
24     0504    0131
25     0502    0119
26     0505    0180
27     0506    0169
28     0506    0128
29     0603    0218
30     0502    0148
..      ...     ...
68     0508    0175
70     0509    0190
72     0506    0150
73     0506    0160
74     0601    0232
75     0508    0222
76     0510    0210
77     0500    0170
78     0500    0237
79     0505    0188
80     0505    0125
81     0506    0180
82     0509    0180
83     0500    0160
84     0507    0225
85     0506    0160
86     0506    0160
87     0500    0120
88     0510    0228
89     0508    0205
91     0600    0235
92     0502    0142
93     0511    0210
94     0509    0182
95     0508    0220
96     0410    0125
97     0600    0190
98     0600    0200
99     0507    0170
100    0510    0200

[95 rows x 2 columns]

In [75]:
#Find rows based on a value
brfss[brfss['WEIGHT2']=='9999']

Empty DataFrame
Columns: [state, imonth, iday, iyear, dispcode, genhlth, physhlth, menthlth, poorhlth, hlthpln1, persdoc2, medcost, checkup1, WEIGHT2, HEIGHT3, wtunit, wt, wtlbs, htunit, htinches]
Index: []

The [`shape()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.shape.html) method gives you the height and width of your DataFrame.

In [76]:
print(brfss.shape)
#Drop any rows without weight
brfss.dropna(subset=['WEIGHT2'], inplace=True)
print(brfss.shape)

(404523, 20)
(404523, 20)


//anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.isin.html

In [77]:
brfss[np.isin(brfss['WEIGHT2'], ['7777', '9999'], invert=True)].shape

(404523, 20)

#### Derived Fields

In [78]:
brfss = brfss[np.isin(brfss['WEIGHT2'], ['7777', '9999'], invert=True)]
# Get rid of one questionable record
brfss = brfss[~brfss.WEIGHT2.str.startswith('1')]

In [79]:
brfss['wtunit'] = brfss.WEIGHT2.str[0].astype(np.uint8)
brfss['wt'] = brfss.WEIGHT2.str[1:].astype(np.uint16)
brfss

state  imonth  iday  iyear  dispcode  genhlth  physhlth  menthlth  \
0           1       1     5   2018      1100      2.0      30.0      88.0   
1           1       1    12   2018      1100      3.0      88.0      88.0   
2           1       1     8   2018      1100      5.0      10.0      88.0   
3           1       1     3   2018      1100      1.0      88.0      88.0   
5           1       1    11   2018      1100      2.0      88.0      88.0   
6           1       1    10   2018      1100      1.0      88.0      88.0   
7           1       1    13   2018      1100      2.0      88.0      88.0   
8           1       1     9   2018      1100      2.0      88.0      88.0   
9           1       1    10   2018      1100      3.0       5.0      88.0   
10          1       1    10   2018      1100      1.0      88.0      88.0   
11          1       1     3   2018      1100      3.0       7.0       2.0   
12          1       1    11   2018      1100      2.0       2.0       1.0   
13          1       1    12   2018      1100      4.0      88.0      88.0   
14          1       1    25   2018      1100      3.0      88.0      88.0   
15          1       1    19   2018      1100      3.0       5.0      88.0   
16          1       1    10   2018      1100      3.0      77.0      88.0   
17          1       1    18   2018      1100      3.0      88.0      88.0   
18          1       1    11   2018      1100      3.0      88.0      88.0   
19          1       1    18   2018      1200      2.0       2.0       2.0   
20          1       1     5   2018      1100      2.0      88.0       3.0   
21          1       1     3   2018      1100      3.0      88.0      30.0   
22          1       1    10   2018      1100      5.0      30.0      88.0   
23          1       1    11   2018      1100      3.0      88.0      88.0   
24          1       1    11   2018      1100      3.0      88.0      88.0   
25          1       1    10   2018      1100      5.0      30.0      88.0   
26          1       1     3   2018      1100      3.0      88.0      88.0   
27          1       1     8   2018      1100      2.0      88.0      88.0   
28          1       1    10   2018      1100      2.0       3.0      88.0   
29          1       1    10   2018      1100      3.0       2.0      15.0   
30          1       1    18   2018      1100      3.0       5.0      88.0   
...       ...     ...   ...    ...       ...      ...       ...       ...   
437404      2       2    28   2019      1100      4.0      30.0      30.0   
437405      2       3     1   2019      1100      2.0      88.0      88.0   
437407      2       3    12   2019      1100      3.0      88.0      88.0   
437408      2       2    25   2019      1100      3.0      88.0      88.0   
437409      2       3    23   2019      1100      4.0      30.0      30.0   
437410      2       3    21   2019      1100      4.0      88.0      88.0   
437411      2       3    23   2019      1100      4.0       3.0      88.0   
437412      2       3     2   2019      1100      3.0       1.0      88.0   
437413      2       2    27   2019      1100      1.0      88.0      88.0   
437414      2       3    22   2019      1100      2.0      88.0      88.0   
437415      2       3    26   2019      1100      2.0      88.0      88.0   
437416      2       2    27   2019      1100      1.0      88.0      15.0   
437417      2       3    10   2019      1100      3.0       2.0      88.0   
437418      2       2    23   2019      1100      3.0       7.0      88.0   
437419      2       2    27   2019      1100      4.0       3.0      88.0   
437420      2       3    12   2019      1100      4.0      30.0      77.0   
437421      2       2    26   2019      1100      3.0       5.0       2.0   
437422      2       2    26   2019      1100      4.0      14.0      10.0   
437423      2       3    12   2019      1100      4.0      88.0      88.0   
437424      2       3    19   2019      1100      4.0      88.0      88.0   
4374

[`np.where()`](https://docs.scipy.org/doc/numpy/reference/generated/numpy.where.html) provides the equivalent of an *if-then-else* statement on each observation in a DataFrame:

In [80]:
lbsperkg = 2.205
brfss['wtlbs'] = np.where(brfss.wtunit==9, brfss.wt*lbsperkg, brfss.wt).astype(np.int16)

In [81]:
brfss[brfss.wtunit==9]

state  imonth  iday  iyear  dispcode  genhlth  physhlth  menthlth  \
2732        1       3    27   2018      1100      3.0      88.0      88.0   
5182        1       9    26   2018      1100      2.0      88.0      88.0   
8442        2       6     4   2018      1100      4.0      30.0      30.0   
9092        2      10    30   2018      1100      2.0       1.0      88.0   
10312       4       8     7   2018      1100      2.0      14.0      88.0   
10404       4      10     4   2018      1100      4.0      88.0      88.0   
10780       4       1    24   2018      1100      3.0      88.0      88.0   
13098       4       2    15   2018      1100      1.0      88.0      88.0   
13124       4       2     7   2018      1100      1.0      88.0      88.0   
13176       4       2    11   2018      1100      1.0      88.0      88.0   
13357       4       3    30   2018      1100      1.0      88.0       3.0   
13380       4       3    18   2018      1100      4.0       3.0      77.0   
13759       4       4    13   2018      1100      1.0      88.0      88.0   
13960       4       7     3   2018      1200      4.0       2.0      88.0   
14365       4       8     8   2018      1100      3.0      88.0      88.0   
14366       4       8     9   2018      1100      2.0      88.0      88.0   
14415       4       8     7   2018      1100      1.0      88.0      88.0   
14599       4      10     1   2018      1100      4.0      15.0      88.0   
14640       4      10    23   2018      1100      4.0       2.0       1.0   
14844       4      10    18   2018      1100      3.0       2.0       1.0   
14896       4      11    19   2018      1100      1.0      88.0      88.0   
15001       4      11     7   2018      1100      5.0      30.0      30.0   
16481       4       7    18   2018      1100      4.0      88.0      88.0   
16905       4       9     8   2018      1100      2.0      88.0      15.0   
16919       4       9    27   2018      1100      2.0      88.0      88.0   
17225       4      10    18   2018      1100      4.0       2.0      88.0   
17392       4      12     8   2018      1200      2.0      88.0      88.0   
24901       6       7    18   2018      1200      3.0      88.0      88.0   
24909       6       9    23   2018      1200      2.0       4.0      88.0   
24954       6      10    31   2018      1200      3.0      88.0      88.0   
...       ...     ...   ...    ...       ...      ...       ...       ...   
422061      5       4    10   2018      1100      2.0      88.0      88.0   
422222      5       5    11   2018      1100      3.0      88.0      88.0   
423440      5       1    11   2019      1200      3.0      88.0      88.0   
424002      5       3     1   2018      1100      1.0      88.0      88.0   
424296      5       3    21   2018      1100      3.0      88.0      88.0   
424329      5       3    26   2018      1200      3.0      88.0      88.0   
424390      5       4    29   2018      1200      2.0       1.0      88.0   
424459      5       4    10   2018      1100      3.0      88.0      88.0   
424990      5       7     8   2018      1100      4.0      14.0      88.0   
425141      5       7    12   2018      1100      4.0      30.0      30.0   
425621      5       9     1   2018      1200      2.0      88.0      10.0   
425814      5       9     8   2018      1100      1.0       1.0      88.0   
426243      5      11    20   2018      1100      3.0      88.0      10.0   
426436      5      12    15   2018      1100      2.0      77.0      88.0   
430311      6       9     9   2018      1100      5.0      30.0      30.0   
430803      6      12     4   2018      1100      2.0      88.0      88.0   
431170      6       4    19   2018      1100      1.0      88.0      88.0   
431206      6       4    19   2018      1100      2.0      88.0      88.0   
431238      6       5    19   2018      1100      4.0       7.0      30.0   
431320      6       6    18   2018      1100      3.0       4.0      88.0   
4313

##### Practice

Create a `htinches` column from the `brfss.HEIGHT3` column. 

1. Remove rows where `HEIGHT3` is 7777 ("Don't know/Not sure"),  9999 ("Refused"), or NaN.
2. If the first character of `HEIGHT3` is '9', multiply the remaining three digits by `cmtoin` (defined below) to get height in inches.
3. If the first character of `HEIGHT3` is '0', the second character is feet, and the third and fourth are inches ('0601' means six feet, one inch). Convert this to inches.

See page 36 of the codebook for details on the `HEIGHT3` field.

In [91]:
cmtoin = 0.3937

In [92]:
brfss.dropna(subset=['HEIGHT3'], inplace=True)

In [93]:
brfss[np.isin(brfss['HEIGHT3'], ['7777', '9999'], invert=True)]

state  imonth  iday  iyear  dispcode  genhlth  physhlth  menthlth  \
0           1       1     5   2018      1100      2.0      30.0      88.0   
1           1       1    12   2018      1100      3.0      88.0      88.0   
2           1       1     8   2018      1100      5.0      10.0      88.0   
3           1       1     3   2018      1100      1.0      88.0      88.0   
5           1       1    11   2018      1100      2.0      88.0      88.0   
6           1       1    10   2018      1100      1.0      88.0      88.0   
7           1       1    13   2018      1100      2.0      88.0      88.0   
8           1       1     9   2018      1100      2.0      88.0      88.0   
9           1       1    10   2018      1100      3.0       5.0      88.0   
10          1       1    10   2018      1100      1.0      88.0      88.0   
11          1       1     3   2018      1100      3.0       7.0       2.0   
12          1       1    11   2018      1100      2.0       2.0       1.0   
13          1       1    12   2018      1100      4.0      88.0      88.0   
14          1       1    25   2018      1100      3.0      88.0      88.0   
15          1       1    19   2018      1100      3.0       5.0      88.0   
16          1       1    10   2018      1100      3.0      77.0      88.0   
17          1       1    18   2018      1100      3.0      88.0      88.0   
18          1       1    11   2018      1100      3.0      88.0      88.0   
19          1       1    18   2018      1200      2.0       2.0       2.0   
20          1       1     5   2018      1100      2.0      88.0       3.0   
21          1       1     3   2018      1100      3.0      88.0      30.0   
22          1       1    10   2018      1100      5.0      30.0      88.0   
23          1       1    11   2018      1100      3.0      88.0      88.0   
24          1       1    11   2018      1100      3.0      88.0      88.0   
25          1       1    10   2018      1100      5.0      30.0      88.0   
26          1       1     3   2018      1100      3.0      88.0      88.0   
27          1       1     8   2018      1100      2.0      88.0      88.0   
28          1       1    10   2018      1100      2.0       3.0      88.0   
29          1       1    10   2018      1100      3.0       2.0      15.0   
30          1       1    18   2018      1100      3.0       5.0      88.0   
...       ...     ...   ...    ...       ...      ...       ...       ...   
437404      2       2    28   2019      1100      4.0      30.0      30.0   
437405      2       3     1   2019      1100      2.0      88.0      88.0   
437407      2       3    12   2019      1100      3.0      88.0      88.0   
437408      2       2    25   2019      1100      3.0      88.0      88.0   
437409      2       3    23   2019      1100      4.0      30.0      30.0   
437410      2       3    21   2019      1100      4.0      88.0      88.0   
437411      2       3    23   2019      1100      4.0       3.0      88.0   
437412      2       3     2   2019      1100      3.0       1.0      88.0   
437413      2       2    27   2019      1100      1.0      88.0      88.0   
437414      2       3    22   2019      1100      2.0      88.0      88.0   
437415      2       3    26   2019      1100      2.0      88.0      88.0   
437416      2       2    27   2019      1100      1.0      88.0      15.0   
437417      2       3    10   2019      1100      3.0       2.0      88.0   
437418      2       2    23   2019      1100      3.0       7.0      88.0   
437419      2       2    27   2019      1100      4.0       3.0      88.0   
437420      2       3    12   2019      1100      4.0      30.0      77.0   
437421      2       2    26   2019      1100      3.0       5.0       2.0   
437422      2       2    26   2019      1100      4.0      14.0      10.0   
437423      2       3    12   2019      1100      4.0      88.0      88.0   
437424      2       3    19   2019      1100      4.0      88.0      88.0   
4374

In [94]:
brfss['htunit'] = brfss.HEIGHT3.str[0].astype(np.uint8)
brfss['height'] = brfss.HEIGHT3.str[1:].astype(np.uint16)
brfss

state  imonth  iday  iyear  dispcode  genhlth  physhlth  menthlth  \
0           1       1     5   2018      1100      2.0      30.0      88.0   
1           1       1    12   2018      1100      3.0      88.0      88.0   
2           1       1     8   2018      1100      5.0      10.0      88.0   
3           1       1     3   2018      1100      1.0      88.0      88.0   
5           1       1    11   2018      1100      2.0      88.0      88.0   
6           1       1    10   2018      1100      1.0      88.0      88.0   
7           1       1    13   2018      1100      2.0      88.0      88.0   
8           1       1     9   2018      1100      2.0      88.0      88.0   
9           1       1    10   2018      1100      3.0       5.0      88.0   
10          1       1    10   2018      1100      1.0      88.0      88.0   
11          1       1     3   2018      1100      3.0       7.0       2.0   
12          1       1    11   2018      1100      2.0       2.0       1.0   
13          1       1    12   2018      1100      4.0      88.0      88.0   
14          1       1    25   2018      1100      3.0      88.0      88.0   
15          1       1    19   2018      1100      3.0       5.0      88.0   
16          1       1    10   2018      1100      3.0      77.0      88.0   
17          1       1    18   2018      1100      3.0      88.0      88.0   
18          1       1    11   2018      1100      3.0      88.0      88.0   
19          1       1    18   2018      1200      2.0       2.0       2.0   
20          1       1     5   2018      1100      2.0      88.0       3.0   
21          1       1     3   2018      1100      3.0      88.0      30.0   
22          1       1    10   2018      1100      5.0      30.0      88.0   
23          1       1    11   2018      1100      3.0      88.0      88.0   
24          1       1    11   2018      1100      3.0      88.0      88.0   
25          1       1    10   2018      1100      5.0      30.0      88.0   
26          1       1     3   2018      1100      3.0      88.0      88.0   
27          1       1     8   2018      1100      2.0      88.0      88.0   
28          1       1    10   2018      1100      2.0       3.0      88.0   
29          1       1    10   2018      1100      3.0       2.0      15.0   
30          1       1    18   2018      1100      3.0       5.0      88.0   
...       ...     ...   ...    ...       ...      ...       ...       ...   
437404      2       2    28   2019      1100      4.0      30.0      30.0   
437405      2       3     1   2019      1100      2.0      88.0      88.0   
437407      2       3    12   2019      1100      3.0      88.0      88.0   
437408      2       2    25   2019      1100      3.0      88.0      88.0   
437409      2       3    23   2019      1100      4.0      30.0      30.0   
437410      2       3    21   2019      1100      4.0      88.0      88.0   
437411      2       3    23   2019      1100      4.0       3.0      88.0   
437412      2       3     2   2019      1100      3.0       1.0      88.0   
437413      2       2    27   2019      1100      1.0      88.0      88.0   
437414      2       3    22   2019      1100      2.0      88.0      88.0   
437415      2       3    26   2019      1100      2.0      88.0      88.0   
437416      2       2    27   2019      1100      1.0      88.0      15.0   
437417      2       3    10   2019      1100      3.0       2.0      88.0   
437418      2       2    23   2019      1100      3.0       7.0      88.0   
437419      2       2    27   2019      1100      4.0       3.0      88.0   
437420      2       3    12   2019      1100      4.0      30.0      77.0   
437421      2       2    26   2019      1100      3.0       5.0       2.0   
437422      2       2    26   2019      1100      4.0      14.0      10.0   
437423      2       3    12   2019      1100      4.0      88.0      88.0   
437424      2       3    19   2019      1100      4.0      88.0      88.0   
4374

In [96]:
brfss['height_inches'] = np.where(brfss.htunit==9, brfss.height*cmtoin, brfss.height).astype(np.int16)

brfss[brfss.htunit==9]

state  imonth  iday  iyear  dispcode  genhlth  physhlth  menthlth  \
2732        1       3    27   2018      1100      3.0      88.0      88.0   
9605        4       1    24   2018      1100      3.0       2.0       2.0   
9740        4       4     5   2018      1100      4.0      15.0      88.0   
10614       4      12     2   2018      1100      3.0      88.0      88.0   
10780       4       1    24   2018      1100      3.0      88.0      88.0   
11434       4       7    10   2018      1100      1.0      88.0      88.0   
12777       4       1    28   2018      1100      2.0      88.0      88.0   
12869       4       2     4   2018      1100      3.0      10.0       2.0   
13124       4       2     7   2018      1100      1.0      88.0      88.0   
13176       4       2    11   2018      1100      1.0      88.0      88.0   
13307       4       2    12   2018      1100      3.0       2.0      88.0   
13380       4       3    18   2018      1100      4.0       3.0      77.0   
13538       4       3    20   2018      1100      4.0      14.0      88.0   
13759       4       4    13   2018      1100      1.0      88.0      88.0   
13960       4       7     3   2018      1200      4.0       2.0      88.0   
13965       4       6    12   2018      1100      3.0      88.0      88.0   
14112       4       7     9   2018      1100      1.0      88.0      88.0   
14251       4       8     8   2018      1100      1.0      88.0      88.0   
14365       4       8     8   2018      1100      3.0      88.0      88.0   
14415       4       8     7   2018      1100      1.0      88.0      88.0   
14599       4      10     1   2018      1100      4.0      15.0      88.0   
14617       4      10    24   2018      1200      4.0       2.0      88.0   
14640       4      10    23   2018      1100      4.0       2.0       1.0   
14834       4      10    20   2018      1100      4.0      88.0      88.0   
14844       4      10    18   2018      1100      3.0       2.0       1.0   
14896       4      11    19   2018      1100      1.0      88.0      88.0   
15647       4       2    12   2018      1100      3.0      88.0      88.0   
15870       4       3    19   2018      1100      3.0      88.0      77.0   
16317       4       6    30   2018      1200      4.0      88.0       7.0   
16481       4       7    18   2018      1100      4.0      88.0      88.0   
...       ...     ...   ...    ...       ...      ...       ...       ...   
416554      3      12    13   2018      1100      3.0      88.0      88.0   
416556      3      12    14   2018      1100      2.0       5.0      88.0   
421203      4      11    18   2018      1100      1.0      30.0       2.0   
422225      5       6     8   2018      1100      3.0      88.0      88.0   
423440      5       1    11   2019      1200      3.0      88.0      88.0   
423867      5       2    11   2018      1100      2.0      10.0      88.0   
423875      5       2    14   2018      1100      2.0      88.0      88.0   
424198      5       3    11   2018      1100      3.0       7.0      88.0   
424459      5       4    10   2018      1100      3.0      88.0      88.0   
425012      5       7     9   2018      1100      3.0      88.0      88.0   
425038      5       6    20   2018      1100      3.0      88.0       3.0   
425063      5       7     7   2018      1200      1.0      88.0      88.0   
425636      5       8    14   2018      1100      5.0      88.0      88.0   
425814      5       9     8   2018      1100      1.0       1.0      88.0   
425901      5       9    29   2018      1200      3.0      88.0      88.0   
425964      5      11     3   2018      1100      2.0       5.0      88.0   
425990      5      11    13   2018      1200      1.0      30.0      30.0   
426010      5      10    17   2018      1100      3.0      88.0      88.0   
426018      5      11     7   2018      1100      5.0      30.0      99.0   
426032      5      10    29   2018      1100      3.0       2.0       4.0   
4260

#### Summary statistics and aggregation

In [98]:
brfss.groupby(['persdoc2', 'poorhlth']).size() #or .mean()

persdoc2  poorhlth
1.0       1.0          7804
          2.0          8895
          3.0          5675
          4.0          3230
          5.0          6204
          6.0          1011
          7.0          3098
          8.0           754
          9.0           174
          10.0         5242
          11.0           60
          12.0          499
          13.0           61
          14.0         1643
          15.0         5461
          16.0          135
          17.0           82
          18.0          160
          19.0           17
          20.0         3538
          21.0          460
          22.0           67
          23.0           44
          24.0           72
          25.0         1357
          26.0           67
          27.0           87
          28.0          274
          29.0          102
          30.0        14046
                      ...  
7.0       14.0            7
          15.0           21
          18.0            1
          20.0           16
 

In [99]:
# | means 'or'
# ph=1 if you were sick more than 5 days, 0 otherwise:
brfss['ph'] = np.where((brfss['poorhlth'] > 30) | (brfss['poorhlth'] <= 5) | (brfss['poorhlth'].isnull()), 0, 1)
brfss

state  imonth  iday  iyear  dispcode  genhlth  physhlth  menthlth  \
0           1       1     5   2018      1100      2.0      30.0      88.0   
1           1       1    12   2018      1100      3.0      88.0      88.0   
2           1       1     8   2018      1100      5.0      10.0      88.0   
3           1       1     3   2018      1100      1.0      88.0      88.0   
5           1       1    11   2018      1100      2.0      88.0      88.0   
6           1       1    10   2018      1100      1.0      88.0      88.0   
7           1       1    13   2018      1100      2.0      88.0      88.0   
8           1       1     9   2018      1100      2.0      88.0      88.0   
9           1       1    10   2018      1100      3.0       5.0      88.0   
10          1       1    10   2018      1100      1.0      88.0      88.0   
11          1       1     3   2018      1100      3.0       7.0       2.0   
12          1       1    11   2018      1100      2.0       2.0       1.0   
13          1       1    12   2018      1100      4.0      88.0      88.0   
14          1       1    25   2018      1100      3.0      88.0      88.0   
15          1       1    19   2018      1100      3.0       5.0      88.0   
16          1       1    10   2018      1100      3.0      77.0      88.0   
17          1       1    18   2018      1100      3.0      88.0      88.0   
18          1       1    11   2018      1100      3.0      88.0      88.0   
19          1       1    18   2018      1200      2.0       2.0       2.0   
20          1       1     5   2018      1100      2.0      88.0       3.0   
21          1       1     3   2018      1100      3.0      88.0      30.0   
22          1       1    10   2018      1100      5.0      30.0      88.0   
23          1       1    11   2018      1100      3.0      88.0      88.0   
24          1       1    11   2018      1100      3.0      88.0      88.0   
25          1       1    10   2018      1100      5.0      30.0      88.0   
26          1       1     3   2018      1100      3.0      88.0      88.0   
27          1       1     8   2018      1100      2.0      88.0      88.0   
28          1       1    10   2018      1100      2.0       3.0      88.0   
29          1       1    10   2018      1100      3.0       2.0      15.0   
30          1       1    18   2018      1100      3.0       5.0      88.0   
...       ...     ...   ...    ...       ...      ...       ...       ...   
437404      2       2    28   2019      1100      4.0      30.0      30.0   
437405      2       3     1   2019      1100      2.0      88.0      88.0   
437407      2       3    12   2019      1100      3.0      88.0      88.0   
437408      2       2    25   2019      1100      3.0      88.0      88.0   
437409      2       3    23   2019      1100      4.0      30.0      30.0   
437410      2       3    21   2019      1100      4.0      88.0      88.0   
437411      2       3    23   2019      1100      4.0       3.0      88.0   
437412      2       3     2   2019      1100      3.0       1.0      88.0   
437413      2       2    27   2019      1100      1.0      88.0      88.0   
437414      2       3    22   2019      1100      2.0      88.0      88.0   
437415      2       3    26   2019      1100      2.0      88.0      88.0   
437416      2       2    27   2019      1100      1.0      88.0      15.0   
437417      2       3    10   2019      1100      3.0       2.0      88.0   
437418      2       2    23   2019      1100      3.0       7.0      88.0   
437419      2       2    27   2019      1100      4.0       3.0      88.0   
437420      2       3    12   2019      1100      4.0      30.0      77.0   
437421      2       2    26   2019      1100      3.0       5.0       2.0   
437422      2       2    26   2019      1100      4.0      14.0      10.0   
437423      2       3    12   2019      1100      4.0      88.0      88.0   
437424      2       3    19   2019      1100      4.0      88.0      88.0   
4374

In [100]:
#brfss.groupby(['persdoc2', 'ph']).describe()
brfss.ph.groupby(brfss.persdoc2).size()

persdoc2
1.0    304377
2.0     31602
3.0     66196
7.0      1243
9.0       302
Name: ph, dtype: int64

In [101]:
#Calculate percentages instead of raw numbers.
docph = brfss.groupby(['persdoc2', 'ph']).size()
docph.groupby(level=0).apply(lambda x: 100 * x / float(x.sum()))

persdoc2  ph
1.0       0     87.347599
          1     12.652401
2.0       0     80.396810
          1     19.603190
3.0       0     89.567345
          1     10.432655
7.0       0     87.610619
          1     12.389381
9.0       0     87.086093
          1     12.913907
dtype: float64

We can bin or categorize numeric variables with [`pd.cut()`](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.cut.html).

In [102]:
## 88 means 'none'; want to bin it separately from "Don't know" and "refused"
brfss.loc[brfss.poorhlth==88, 'poorhlth']=51
bins = [1, 5, 10, 20, 30, 51, 70]
brfss['phcats'] = pd.cut(brfss.poorhlth, bins, 
                         labels=['less than 5', 'less than 10', 
                                 'less than 20', 'more than 20', 'none', "don't know/refused"])
brfss

state  imonth  iday  iyear  dispcode  genhlth  physhlth  menthlth  \
0           1       1     5   2018      1100      2.0      30.0      88.0   
1           1       1    12   2018      1100      3.0      88.0      88.0   
2           1       1     8   2018      1100      5.0      10.0      88.0   
3           1       1     3   2018      1100      1.0      88.0      88.0   
5           1       1    11   2018      1100      2.0      88.0      88.0   
6           1       1    10   2018      1100      1.0      88.0      88.0   
7           1       1    13   2018      1100      2.0      88.0      88.0   
8           1       1     9   2018      1100      2.0      88.0      88.0   
9           1       1    10   2018      1100      3.0       5.0      88.0   
10          1       1    10   2018      1100      1.0      88.0      88.0   
11          1       1     3   2018      1100      3.0       7.0       2.0   
12          1       1    11   2018      1100      2.0       2.0       1.0   
13          1       1    12   2018      1100      4.0      88.0      88.0   
14          1       1    25   2018      1100      3.0      88.0      88.0   
15          1       1    19   2018      1100      3.0       5.0      88.0   
16          1       1    10   2018      1100      3.0      77.0      88.0   
17          1       1    18   2018      1100      3.0      88.0      88.0   
18          1       1    11   2018      1100      3.0      88.0      88.0   
19          1       1    18   2018      1200      2.0       2.0       2.0   
20          1       1     5   2018      1100      2.0      88.0       3.0   
21          1       1     3   2018      1100      3.0      88.0      30.0   
22          1       1    10   2018      1100      5.0      30.0      88.0   
23          1       1    11   2018      1100      3.0      88.0      88.0   
24          1       1    11   2018      1100      3.0      88.0      88.0   
25          1       1    10   2018      1100      5.0      30.0      88.0   
26          1       1     3   2018      1100      3.0      88.0      88.0   
27          1       1     8   2018      1100      2.0      88.0      88.0   
28          1       1    10   2018      1100      2.0       3.0      88.0   
29          1       1    10   2018      1100      3.0       2.0      15.0   
30          1       1    18   2018      1100      3.0       5.0      88.0   
...       ...     ...   ...    ...       ...      ...       ...       ...   
437404      2       2    28   2019      1100      4.0      30.0      30.0   
437405      2       3     1   2019      1100      2.0      88.0      88.0   
437407      2       3    12   2019      1100      3.0      88.0      88.0   
437408      2       2    25   2019      1100      3.0      88.0      88.0   
437409      2       3    23   2019      1100      4.0      30.0      30.0   
437410      2       3    21   2019      1100      4.0      88.0      88.0   
437411      2       3    23   2019      1100      4.0       3.0      88.0   
437412      2       3     2   2019      1100      3.0       1.0      88.0   
437413      2       2    27   2019      1100      1.0      88.0      88.0   
437414      2       3    22   2019      1100      2.0      88.0      88.0   
437415      2       3    26   2019      1100      2.0      88.0      88.0   
437416      2       2    27   2019      1100      1.0      88.0      15.0   
437417      2       3    10   2019      1100      3.0       2.0      88.0   
437418      2       2    23   2019      1100      3.0       7.0      88.0   
437419      2       2    27   2019      1100      4.0       3.0      88.0   
437420      2       3    12   2019      1100      4.0      30.0      77.0   
437421      2       2    26   2019      1100      3.0       5.0       2.0   
437422      2       2    26   2019      1100      4.0      14.0      10.0   
437423      2       3    12   2019      1100      4.0      88.0      88.0   
437424      2       3    19   2019      1100      4.0      88.0      88.0   
4374

##### Practice

Create a new column that divides `iday` into 3 bins (1-10, 11-20, 21+).

In [104]:
## 88 means 'none'; want to bin it separately from "Don't know" and "refused"
bins2 = [1, 10, 20, 30]
brfss['daycats'] = pd.cut(brfss.iday, bins2, 
                         labels=['less than 10', 'less than 20', 
                                 'more than 20'])
brfss

state  imonth  iday  iyear  dispcode  genhlth  physhlth  menthlth  \
0           1       1     5   2018      1100      2.0      30.0      88.0   
1           1       1    12   2018      1100      3.0      88.0      88.0   
2           1       1     8   2018      1100      5.0      10.0      88.0   
3           1       1     3   2018      1100      1.0      88.0      88.0   
5           1       1    11   2018      1100      2.0      88.0      88.0   
6           1       1    10   2018      1100      1.0      88.0      88.0   
7           1       1    13   2018      1100      2.0      88.0      88.0   
8           1       1     9   2018      1100      2.0      88.0      88.0   
9           1       1    10   2018      1100      3.0       5.0      88.0   
10          1       1    10   2018      1100      1.0      88.0      88.0   
11          1       1     3   2018      1100      3.0       7.0       2.0   
12          1       1    11   2018      1100      2.0       2.0       1.0   
13          1       1    12   2018      1100      4.0      88.0      88.0   
14          1       1    25   2018      1100      3.0      88.0      88.0   
15          1       1    19   2018      1100      3.0       5.0      88.0   
16          1       1    10   2018      1100      3.0      77.0      88.0   
17          1       1    18   2018      1100      3.0      88.0      88.0   
18          1       1    11   2018      1100      3.0      88.0      88.0   
19          1       1    18   2018      1200      2.0       2.0       2.0   
20          1       1     5   2018      1100      2.0      88.0       3.0   
21          1       1     3   2018      1100      3.0      88.0      30.0   
22          1       1    10   2018      1100      5.0      30.0      88.0   
23          1       1    11   2018      1100      3.0      88.0      88.0   
24          1       1    11   2018      1100      3.0      88.0      88.0   
25          1       1    10   2018      1100      5.0      30.0      88.0   
26          1       1     3   2018      1100      3.0      88.0      88.0   
27          1       1     8   2018      1100      2.0      88.0      88.0   
28          1       1    10   2018      1100      2.0       3.0      88.0   
29          1       1    10   2018      1100      3.0       2.0      15.0   
30          1       1    18   2018      1100      3.0       5.0      88.0   
...       ...     ...   ...    ...       ...      ...       ...       ...   
437404      2       2    28   2019      1100      4.0      30.0      30.0   
437405      2       3     1   2019      1100      2.0      88.0      88.0   
437407      2       3    12   2019      1100      3.0      88.0      88.0   
437408      2       2    25   2019      1100      3.0      88.0      88.0   
437409      2       3    23   2019      1100      4.0      30.0      30.0   
437410      2       3    21   2019      1100      4.0      88.0      88.0   
437411      2       3    23   2019      1100      4.0       3.0      88.0   
437412      2       3     2   2019      1100      3.0       1.0      88.0   
437413      2       2    27   2019      1100      1.0      88.0      88.0   
437414      2       3    22   2019      1100      2.0      88.0      88.0   
437415      2       3    26   2019      1100      2.0      88.0      88.0   
437416      2       2    27   2019      1100      1.0      88.0      15.0   
437417      2       3    10   2019      1100      3.0       2.0      88.0   
437418      2       2    23   2019      1100      3.0       7.0      88.0   
437419      2       2    27   2019      1100      4.0       3.0      88.0   
437420      2       3    12   2019      1100      4.0      30.0      77.0   
437421      2       2    26   2019      1100      3.0       5.0       2.0   
437422      2       2    26   2019      1100      4.0      14.0      10.0   
437423      2       3    12   2019      1100      4.0      88.0      88.0   
437424      2       3    19   2019      1100      4.0      88.0      88.0   
4374

2) Group `brfss` by `hlthpln1` and `medcost` and create a table like the one above (for `persdoc2` and `ph`) with percentages for each subgroup.

In [105]:
brfss.hlthpln1.groupby(brfss.medcost).size()

medcost
1.0     41114
2.0    361626
7.0       819
9.0       160
Name: hlthpln1, dtype: int64

In [106]:
hlthmedcost = brfss.groupby(['hlthpln1', 'medcost']).size()
hlthmedcost.groupby(level=0).apply(lambda x: 100 * x / float(x.sum()))

hlthpln1  medcost
1.0       1.0         7.947766
          2.0        91.840247
          7.0         0.178586
          9.0         0.033401
2.0       1.0        36.722233
          2.0        62.793177
          7.0         0.423209
          9.0         0.061381
7.0       1.0        17.809524
          2.0        79.904762
          7.0         2.190476
          9.0         0.095238
9.0       1.0        11.587983
          2.0        84.549356
          7.0         0.429185
          9.0         3.433476
dtype: float64